pkg/extensions/extension_scrub.go

```diff
@@ -4,30 +4,107 @@
-func EnableScrubExtension(config *config.Config, log log.Logger, run bool, imgStore storage.ImageStore, repo string) {
-	if !run {
-		if config.Extensions.Scrub != nil &&
-			config.Extensions.Scrub.Interval != 0 {
-			minScrubInterval, _ := time.ParseDuration("2h")
+func EnableScrubExtension(config *config.Config, log log.Logger, storeController storage.StoreController,
+	sch *scheduler.TaskScheduler,
+) {
+	if config.Extensions.Scrub != nil &&
+		config.Extensions.Scrub.Interval != 0 {
+		minScrubInterval, _ := time.ParseDuration("2h")
 
-			if config.Extensions.Scrub.Interval < minScrubInterval {
-				config.Extensions.Scrub.Interval = minScrubInterval
+		if config.Extensions.Scrub.Interval < minScrubInterval {
+			config.Extensions.Scrub.Interval = minScrubInterval
 
-				log.Warn().Msg("Scrub interval set to too-short interval < 2h, changing scrub duration to 2 hours and continuing.") //nolint:lll // gofumpt conflicts with lll
+			log.Warn().Msg("Scrub interval set to too-short interval < 2h, changing scrub duration to 2 hours and continuing.") //nolint:lll // gofumpt conflicts with lll
+		}
+
+		var periodicScrub *periodicScrubTask
+
+		periodicScrub = newPeriodicScrubTask(config.Extensions.Scrub.Interval, storeController.DefaultStore, log, sch)
+		sch.AddTask(periodicScrub)
+
+		if config.Storage.SubPaths != nil {
+			for route := range config.Storage.SubPaths {
+				periodicScrub = newPeriodicScrubTask(config.Extensions.Scrub.Interval, storeController.SubStore[route], log, sch)
+				sch.AddTask(periodicScrub)
 			}
-		} else {
-			log.Info().Msg("Scrub config not provided, skipping scrub")
 		}
 	} else {
-		scrub.RunScrubRepo(imgStore, repo, log)
+		log.Info().Msg("Scrub config not provided, skipping scrub")
 	}
 }
+
+type periodicScrubTask struct {
+	*scheduler.AbstractTask
+	interval time.Duration
+	imgStore storage.ImageStore
+	log      log.Logger
+	sch      *scheduler.TaskScheduler
+	cancel   chan bool
+}
+
+func newPeriodicScrubTask(interval time.Duration, imgStore storage.ImageStore, log log.Logger,
+	sch *scheduler.TaskScheduler,
+) *periodicScrubTask {
+	abstractTask := &scheduler.AbstractTask{}
+	cancel := make(chan bool)
+	periodicScrub := &periodicScrubTask{abstractTask, interval, imgStore, log, sch, cancel}
+	abstractTask.Task = periodicScrub
+
+	return periodicScrub
+}
+
+func (periodicScrubT *periodicScrubTask) DoWork(wgScheduler *goSync.WaitGroup) (scheduler.Status, scheduler.Cb) {
+	ticker := time.NewTicker(periodicScrubT.interval)
+
+	go func() {
+		for {
+			select {
+			case <-periodicScrubT.cancel:
+				ticker.Stop()
+
+				if wgScheduler != nil {
+					wgScheduler.Done()
+				}
+
+				return
+			default:
+				repo, errIterator := periodicScrubT.imgStore.GetFirstRepository()
+
+				for {
+					if errIterator != nil && !errors.Is(errIterator, io.EOF) {
+						break
+					}
+
+					if repo == "" {
+						break
+					}
+
+					scrubTask := scrub.NewScrubTask(periodicScrubT.imgStore, repo, periodicScrubT.log)
```

Comment:
Can we have the task DoWork in a generic method, and the scrub logic be passed as a parameter of type function to it? Most of this method looks like code which would be duplicated in all DoWork implementations.

We would pass the actual "work"/scrub function as well as the callback function(s ?) for success or failure to the generic DoWork function. These work/context functions would be called using the same context as parameter (which would be instantiated just before calling DoWork for this specific task and passed on as a parameter), so if one is cancelled, all are cancelled.